kv.go

```diff
@@ -0,0 +1,488 @@
+	}
+	return err
+}
+
+// History will return all values for the key.
+func (kv *kvs) History(key string) ([]KeyValueEntry, error) {
+	// Do quick check to make sure this is a legit key with history.
+	if _, err := kv.Get(key); err != nil && err != ErrKeyDeleted {
+		return nil, err
+	}
+
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return nil, err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	ctx := o.ctx
+	if ctx == nil {
+		ctx, cancel = context.WithTimeout(context.Background(), o.wait)
+		defer cancel()
+	}
+
+	var vals []KeyValueEntry
+	done := make(chan error, 1)
+	cb := func(m *Msg) {
+		tokens, err := getMetadataFields(m.Reply)
+		if err != nil {
+			done <- err
+		} else {
+			if len(m.Subject) <= len(kv.pre) {
+				done <- ErrBadSubject
+				return
+			}
+			subj := m.Subject[len(kv.pre):]
+			pending := tokens[ackNumPendingTokenPos]
+			vals = append(vals, &kve{
+				bucket:   kv.name,
+				key:      subj,
+				value:    m.Data,
+				revision: uint64(parseNum(tokens[ackStreamSeqTokenPos])),
+				created:  time.Unix(0, parseNum(tokens[ackTimestampSeqTokenPos])),
+				delta:    uint64(parseNum(pending)),
+			})
+			if pending == kvNoPending {
+				done <- nil
+			}
+		}
+	}
+
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	// Used ordered consumer to deliver results.
+	sub, err := kv.js.Subscribe(b.String(), cb, OrderedConsumer())
+	if err != nil {
+		return nil, err
+	}
+	defer sub.Unsubscribe()
+
+	// Wait on done or ctx/timeout.
+	select {
+	case err := <-done:
+		if err != nil {
+			return nil, err
+		}
+		return vals, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// WatchAll watches all keys.
+func (kv *kvs) WatchAll(cb KeyValueUpdate) (*Subscription, error) {
+	return kv.Watch(AllKeys, cb)
+}
+
+// Watch will fire the callback when a key that matches the keys pattern is updated.
+// keys needs to be a valid NATS subject.
+func (kv *kvs) Watch(keys string, cb KeyValueUpdate) (*Subscription, error) {
+	// Could be a pattern so don't check for validity as we normally do.
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(keys)
+
+	update := func(m *Msg) {
+		tokens, err := getMetadataFields(m.Reply)
+		if err != nil {
+			return
+		}
+		if len(m.Subject) <= len(kv.pre) {
+			//done <- ErrBadSubject
+			return
+		}
+		subj := m.Subject[len(kv.pre):]
+
+		var op KeyValueOp
+		if len(m.Header) > 0 && m.Header.Get(kvop) == kvdel {
+			op = KeyValueDelete
+		}
+
+		cb(&kve{
+			bucket:   kv.name,
+			key:      subj,
+			value:    m.Data,
+			revision: uint64(parseNum(tokens[ackStreamSeqTokenPos])),
+			created:  time.Unix(0, parseNum(tokens[ackTimestampSeqTokenPos])),
+			delta:    uint64(parseNum(tokens[ackNumPendingTokenPos])),
+			op:       op,
+		})
+	}
+	// Used ordered consumer to deliver results.
+	return kv.js.Subscribe(b.String(), update, OrderedConsumer(), DeliverLastPerSubject())
```

Comment:
There's a missing behavior here mentioned in the ADR:

```
When a stream is empty at create time - `NumPending+Delivered.Consumer==0` - a nil is sent across the channel to signal to the caller that they should not expect any values.
```

The purpose of this is in creating in-memory caches, you want the cache to be ready and usable as soon as possible, generally the cache can figure this out by looking at delta in the received entries (delta==0 is last message).

But when there are no messages at all in the stream the watch will appear to just wait, the calling code will never be notified that there is no waiting messages and that its in effect ready even when no entries came out of the watch.

Without this a cache would never know if its ready - or perhaps just disconnected or have a broken path between it and the stream

Reply:
not at the moment required and it would be an opt in for using the KV not on by default. We're focussing on getting core behaviors right now.

But we need to make it possible to build such a cache and I have a implementation that I use to test that its possible and makes sense.

Reply:
@derekcollison Is ordered consumer necessary here? If someone wants to audit and not miss an entry, then fine, but if I'm building a cache and miss an entry but get the latest, what's the harm? Maybe ordered should be an option to the watch.

Reply:
Building a cache and missing an entry is pretty serious, when that entry is a delete.

Your cache will not be consistent with reality.

Reply:
Also about a local cache. Is it the expectation that when a KV context is created and a local cache is started, that we warm the entire cache by reading everything in the kv? It seems that it's sufficient to watch and use the results of get, and even expensive to try to warm a cache, when a user wants a cache, again there certainly could be reasons for a client to not want a cache.

Reply:
Cache design not final, the implementation I have do warm itself and acts as a pass through when not ready, but a simple on demand read cache is also of course possible.

A cache built with a watcher must be able to delete items when they are deleted / purged elsewhere though, hence pointing out the missing behavior here.

Reply:
For cache it should not be considered to be 1:1 match for system of record, so what I mean is cache is ready even before any callbacks or populations happen since if you miss (no cache record) you have to ask system of record anyway.

So not sure we need the empty marker, WDYT?

Reply:
No, I 100% think we need the marker. You can make a simple read cache but one populated by a watcher is slightly different and the marker will help in various scenarios.

You can trust me, I've written several of these caches and the feature was added to the ADR since its needed

Reply:
I have written quite a few as well so maybe not understanding. What does the boolean tell you besides that you have received all first (last) values?

What I mean is that the delete markers will populate, say you don't wait (which I am suggesting we should not) when you hook the watcher. A Get that does not see any information in the cache, no data or delete marker, has to be passed to system of record.

So if you wait you might save the first Get that would miss, maybe.. But nothing else imo but happy to understand more here.

Reply:
when you start the watcher on an empty stream, its not going to get anything, it will never call the callback until the first write to the bucket happens.

From the perspective of the code that started the watch, what does it mean when that nothing is happening?

 * Is there no data?
 * Is the ordered consumer disconnected and trying to connect again?
 * Did the bucket vanish and it's repeatedly trying to recreate the consumers?

With the marker saying "there is no data" the calling code can know it can go ahead and do whatever it felt like doing with the data it gathers from the watch.

This marker only happens on EMPTY buckets.

Reply:
And remember, it's not always that easy in watch we can do watch `x.>`, you cannot do a Get() trick to find out the empty stream scenario. 

The only way to do this is with the method described, users would have no way to build something as basic as "dump all values under x.>" without having signal indicating that there's no data matching the pattern.

Reply:
After speaking directly with @ripienaar I understand better what he was aiming at, will add it in with Keys() as well.